octoprint_octobullet/__init__.py

```diff
@@ -29,10 +33,106 @@ def __init__(self):
 		self._bullet = None
 		self._channel = None
 		self._sender = None
+		self._pc_interval = 5
+		self._message_count = 0
+		self._quiet_time_sec = 1800
+		self._last_message = 0
+		self._etl_format   = "{days:d}:{hours:02d}h {minutes:02d}m"
+		self._eta_strftime = "%H:%M %d-%m"
 
 	def _connect_bullet(self, apikey, channel_name=""):
 		self._bullet, self._sender = self._create_sender(apikey, channel_name)
 
+	#~~ progress message helpers
+	def _sanitize_current_data(self, currentData):
+		if (currentData["progress"]["printTimeLeft"] == None):
+			currentData["progress"]["printTimeLeft"] = currentData["job"]["estimatedPrintTime"]
+		if (currentData["progress"]["printTimeLeft"] == None):
+			self._logger.info("Messasge: Still got no print time {}".format(currentData["progress"]["printTimeLeft"]))
+			currentData["progress"]["printTimeLeft"] = 1000
+		if (currentData["progress"]["filepos"] == None):
+			currentData["progress"]["filepos"] = 0
+		if (currentData["progress"]["printTime"] == None):
+			currentData["progress"]["printTime"] = currentData["job"]["estimatedPrintTime"]
+
+		currentData["progress"]["printTimeLeftString"] = "No ETL yet"
+		currentData["progress"]["printTimeString"] = "Not started yet"
+		currentData["progress"]["ETA"] = "No ETA yet"
+		#Add additional data
+		try:
+			currentData["progress"]["printTimeLeftString"] = self._get_time_from_seconds(currentData["progress"]["printTimeLeft"])
+			currentData["progress"]["printTimeString"] = self._get_time_from_seconds(currentData["progress"]["printTime"])
+			currentData["progress"]["ETA"] = time.strftime(self._eta_strftime, time.localtime(time.time() + currentData["progress"]["printTimeLeft"]))
+		except Exception as e:
+			self._logger.warning("Caught an exception trying to parse data: {0}\n Error is: {1}\nTraceback:{2}".format(currentData,e,traceback.format_exc()))
+			
+		return currentData
+
+	def _get_time_from_seconds(self, seconds):
+		days =0
+		hours = 0
+		minutes = 0
+		if seconds >= 86400:
+			days = int(seconds / 86400)
+			secionds = seconds % 86400
+		if seconds >= 3600:
+			hours = int(seconds / 3600)
+			seconds = seconds % 3600
+		if seconds >= 60:
+			minutes = int(seconds / 60)
+			seconds = seconds % 60
+		return self._etl_format.format(**locals())
+	
+	#~~ PrintProgressPlugin
+	def on_print_progress(self, storage, path, progress):
```

Comment:
I have to admit I don't fully understand the calculations going on in this method, but I have the feeling it is way too complicated for what it's supposed to achieve (regularly send an update to the user).

From what I understand your calculations are supposed to limit the amount of messages sent out due to the API limitations of the Pushbullet API, however, it's completely intransparent to me HOW it's supposed to do that and also it appears to do it in a very convulted way. Since in the end I'll be responsible of maintaining this, some explanation would be great :)

What I'd prefer here is splitting this method up to make operation a bit clearer. First of all determine if according to rate limiting you want to send a message in the first place. If not, abort and return. Once you've figured out you want to send a message, fetch the data you need for that, format it and go for it. Replace values you don't know yet with a placeholder ("unknown"?) instead of arbitrarily determined values. Also just fetch what you need from the data structure provided by the backend once instead of multiple times and create your own data structure. Makes things also way more readable due to less [][][] noise.